content/interoperability.tex

```diff
@@ -0,0 +1,191 @@
+    initialization call of the other model in the same program.
+
+    \item The \VAR{THREAD\_FUNNELED} thread level allows only the main thread to
+    make communication calls. A hybrid program using the \VAR{THREAD\_FUNNELED}
+    thread level in both \openshmem and MPI should ensure that the same main thread
+    is used in both communication environments.
+
+    \item The \VAR{THREAD\_SERIALIZED} thread level requires the program to ensure
+    that communication calls are not made concurrently by multiple threads. If a
+    hybrid program uses \VAR{THREAD\_SERIALIZED} in one communication environment
+    and \VAR{THREAD\_SERIALIZED} or \VAR{THREAD\_FUNNELED} in the other one, it
+    should also guarantee that the \openshmem and MPI calls are not made concurrently
+    from two distinct threads.
+\end{itemize}
+
+\subsection{Mapping Process Identification Numbers}
+\label{subsec:interoperability:id}
+
+Similar to the PE identifier in \openshmem, MPI defines rank as the
+identification number of a process in a communicator. Both the \openshmem PE
+and the MPI rank are unique integers assigned from zero to one less than the total
+number of processes. In a hybrid program, the \openshmem
+PE and the MPI rank in \VAR{MPI\_COMM\_WORLD} of a process can be equal.
+This feature, however, may be provided by only some of the \openshmem and MPI
+implementations (e.g., if both environments share the same underlying process
+manager) and is not portably guaranteed. A portable program should always
+use the standard functions in each model, namely, \FUNC{shmem\_my\_pe} in \openshmem
+and \FUNC{MPI\_Comm\_rank} in MPI, to query the process identification numbers
+in each communication environment and manage the mapping of identifiers in the
+program when necessary.
+
+\subsubsection*{Example}
+\label{subsubsec:interoperability:id:example}
+The following example demonstrates how to manage the mapping between \openshmem
+PE identifier and MPI ranks in \VAR{MPI\_COMM\_WORLD} in a hybrid \openshmem
+and MPI program.
+
+\lstinputlisting[language={C}, tabsize=2,
+      basicstyle=\ttfamily\footnotesize]
+      {example_code/hybrid_mpi_mapping_id.c}
+
+\subsection{RMA Programming Models}
+\label{subsec:interoperability:rma}
+
+\openshmem and MPI each define similar one-sided communication models;
+however, a portable program should not assume interoperability between these
+models.
+For instance, \openshmem guarantees the atomicity only of concurrent \openshmem AMO operations
+that operate on symmetric data with the same datatype. Access to the same symmetric
+object with MPI atomic operations, such as an \FUNC{MPI\_Fetch\_and\_op}, may
+result in an undefined result. Furthermore,
+because most RMA programs can be written by using either \openshmem or MPI RMA,
+users should choose only one of the RMA models in the same program, whenever
+possible, for performance and code simplicity.
```

Comment:
If a user is using a library that uses MPI RMA, they may not have an either-or choice.  Phase-based access (e.g. MPI I/O and SHMEM RMA phases) using the different models should still be fine.  Suggest instead:
```
Users should avoid situations where MPI and \openshmem operations perform concurrent
accesses the same memory locations.
```

Reply:
Makes sense.